qiskit_experiments/data_processing/nodes.py

```diff
@@ -56,130 +106,266 @@ def _format_data(self, datum: Any) -> Any:
+                raise DataProcessorError(
+                    f"IQ data error given to {self.__class__.__name__} must be an N dimensional"
+                    f"array with N in {self._required_dimension()}. Instead, a {len(error.shape)}D"
+                    f" array was given."
+                )
+
+            if error is not None and len(error.shape) != len(datum.shape):
+                raise DataProcessorError(
+                    "Datum and error do not have the same shape: "
+                    f"{len(datum.shape)} != {len(error.shape)}."
+                )
+
+        return datum, error
 
     def __repr__(self):
         """String representation of the node."""
         return f"{self.__class__.__name__}(validate: {self._validate}, scale: {self.scale})"
 
 
-class ToReal(IQPart):
-    """IQ data post-processing. Isolate the real part of single-shot IQ data."""
+class SVD(IQPart):
+    """Singular Value Decomposition of averaged IQ data."""
 
-    def _required_dimension(self) -> int:
-        """Require memory to be a 3D array."""
-        return 3
+    def __init__(self, validate: bool = True):
+        """
+        Args:
+            validate: If set to False the DataAction will not validate its input.
+        """
+        super().__init__(validate=validate)
+        self._main_axes = None
+        self._means = None
+        self._scales = None
 
-    def _process(self, datum: np.array) -> np.array:
-        """Take the real part of the IQ data.
+    def _required_dimension(self) -> Set[int]:
+        """Require memory to be a 2D array."""
+        return {2}
+
+    @property
+    def axis(self) -> List[np.array]:
+        """Return the axis of the trained SVD"""
+        if self._main_axes:
+            return self._main_axes
+
+        raise DataProcessorError("SVD is not trained.")
+
+    def means(self, qubit: int, iq_index: int) -> float:
+        """Return the mean by which to correct the IQ data.
+
+        Before training the SVD the mean of the training data is subtracted from the
+        training data to avoid large offsets in the data. These means can be retrieved
+        with this function.
 
         Args:
-            datum: A 3D array of shots, qubits, and a complex IQ point as [real, imaginary].
+            qubit: Index of the qubit.
+            iq_index: Index of either the in-phase (i.e. 0) or the quadrature (i.e. 1).
 
         Returns:
-            A 2D array of shots, qubits. Each entry is the real part of the given IQ data.
+            The mean that was determined during training for the given qubit and IQ index.
         """
-        if self.scale is None:
-            return datum[:, :, 0]
+        return self._means[qubit][iq_index]
 
-        return datum[:, :, 0] * self.scale
+    @property
+    def scales(self) -> List[float]:
+        """Return the scaling of the SVD."""
+        if self._scales:
+            return self._scales
```

Comment:
This sounds like returning list of 1 if not assigned.

Reply:
The rational for using the property is so that users can see the outcome of the training for the purpose of figuring out what happened to the data if need be so I would rather not return a list of ones. With that in mind raising an error is probably too brutal of a behavior. See here: abbce07